File: java/src/test/java/org/libpetri/smt/encoding/NetFlattenerTest.java
package org.libpetri.smt.encoding;

import org.libpetri.analysis.EnvironmentAnalysisMode;
import org.libpetri.core.*;
import org.libpetri.fixtures.PaperNetworks;
import org.junit.jupiter.api.Test;

import java.util.Set;

import static org.junit.jupiter.api.Assertions.*;

class NetFlattenerTest {

    @Test
    void flattenBasicTpn_assignsStablePlaceIndices() {
        var net = PaperNetworks.createBasicTpn();
        var flatNet = NetFlattener.flatten(net, Set.of(), EnvironmentAnalysisMode.ignore());

        // Places should be sorted by name
        for (int i = 1; i < flatNet.places().size(); i++) {
            assertTrue(
                flatNet.places().get(i - 1).name().compareTo(flatNet.places().get(i).name()) <= 0,
                "Places should be sorted by name"
            );
        }

        // Reverse lookup should be consistent
        for (int i = 0; i < flatNet.places().size(); i++) {
            assertEquals(i, flatNet.indexOf(flatNet.places().get(i)));
        }
    }

    @Test
    void flattenBasicTpn_correctPlaceAndTransitionCount() {
        var net = PaperNetworks.createBasicTpn();
        var flatNet = NetFlattener.flatten(net, Set.of(), EnvironmentAnalysisMode.ignore());

        // BasicTPN has 9 places: Pending, Ready, Validated, Understood, Informed, Promoted, Found, Drafted, Answered
        // (net.places() may be incomplete for new-API transitions, but flatten collects all)
        assertEquals(9, flatNet.placeCount());
        // BasicTPN has no XOR, so transitions should be 1:1 = 7
        assertEquals(7, flatNet.transitionCount());
    }

    @Test
    void flattenExtendedTpn_expandsXorBranches() {
        var net = PaperNetworks.createExtendedTpn();
        var flatNet = NetFlattener.flatten(net, Set.of(), EnvironmentAnalysisMode.ignore());

        // ExtendedTPN has 2 XOR transitions (Search, Compose), each with 2 branches
        // So we should have 11 original + 2 extra = 13 flat transitions
        // (Search -> 2 branches, Compose -> 2 branches, rest are 1:1)
        int xorExpansion = 0;
        for (var ft : flatNet.transitions()) {
            if (ft.branchIndex() >= 0) {
                xorExpansion++;
            }
        }
        // Search has 2 branches, Compose has 2 branches = 4 total expanded
        assertEquals(4, xorExpansion, "Should have 4 XOR-expanded transitions");
    }

    @Test
    void flattenTransition_preVectorCorrect() {
        var p1 = Place.of("A", String.class);
        var p2 = Place.of("B", String.class);
        var p3 = Place.of("C", String.class);

        var t = Transition.builder("T")
            .inputs(In.one(p1), In.exactly(3, p2))
            .outputs(Out.place(p3))
            .build();

        var net = PetriNet.builder("Test").transitions(t).build();
        var flatNet = NetFlattener.flatten(net, Set.of(), EnvironmentAnalysisMode.ignore());

        assertEquals(1, flatNet.transitionCount());
        var ft = flatNet.transitions().getFirst();

        int idxA = flatNet.indexOf(p1);
        int idxB = flatNet.indexOf(p2);
        int idxC = flatNet.indexOf(p3);

        assertEquals(1, ft.preVector()[idxA]);
        assertEquals(3, ft.preVector()[idxB]);
        assertEquals(0, ft.preVector()[idxC]);
        assertEquals(0, ft.postVector()[idxA]);
        assertEquals(0, ft.postVector()[idxB]);
        assertEquals(1, ft.postVector()[idxC]);
    }

    @Test
    void flattenTransition_consumeAllFlagSet() {
        var p1 = Place.of("A", String.class);
        var p2 = Place.of("B", String.class);

        var t = Transition.builder("T")
            .inputs(In.all(p1))
            .outputs(Out.place(p2))
            .build();

        var net = PetriNet.builder("Test").transitions(t).build();
        var flatNet = NetFlattener.flatten(net, Set.of(), EnvironmentAnalysisMode.ignore());

        var ft = flatNet.transitions().getFirst();
        int idxA = flatNet.indexOf(p1);
        assertTrue(ft.consumeAll()[idxA], "consumeAll should be true for In.all()");
        assertEquals(1, ft.preVector()[idxA], "pre should be 1 (minimum) for In.all()");
    }

    @Test
    void flattenTransition_inhibitorAndReadArcs() {
        var p1 = Place.of("A", String.class);
        var p2 = Place.of("B", String.class);
        var p3 = Place.of("C", String.class);
        var p4 = Place.of("D", String.class);

        var t = Transition.builder("T")
            .inputs(In.one(p1))
            .inhibitor(p2)
            .read(p3)
            .outputs(Out.place(p4))
            .build();

        var net = PetriNet.builder("Test").transitions(t).build();
        var flatNet = NetFlattener.flatten(net, Set.of(), EnvironmentAnalysisMode.ignore());

        var ft = flatNet.transitions().getFirst();
        int idxB = flatNet.indexOf(p2);
        int idxC = flatNet.indexOf(p3);

        assertArrayEquals(new int[]{idxB}, ft.inhibitorPlaces());
        assertArrayEquals(new int[]{idxC}, ft.readPlaces());
    }

    @Test
    void flattenWithEnvironmentBounds() {
        var envPlace = Place.of("Env", String.class);
        var env = EnvironmentPlace.of(envPlace);
        var output = Place.of("Out", String.class);

        var t = Transition.builder("T")
            .inputs(In.one(envPlace))
            .outputs(Out.place(output))
            .build();

        var net = PetriNet.builder("Test").transitions(t).build();
        var flatNet = NetFlattener.flatten(net, Set.of(env), EnvironmentAnalysisMode.bounded(3));

        assertEquals(3, flatNet.environmentBounds().get(envPlace));
    }

    @Test
    void flattenXorTransition_createsSeparateFlatTransitions() {
        var input = Place.of("In", String.class);
        var out1 = Place.of("Out1", String.class);
        var out2 = Place.of("Out2", String.class);

        var t = Transition.builder("Choice")
            .inputs(In.one(input))
            .outputs(Out.xor(out1, out2))
            .build();

        var net = PetriNet.builder("Test").transitions(t).build();
        var flatNet = NetFlattener.flatten(net, Set.of(), EnvironmentAnalysisMode.ignore());

        assertEquals(2, flatNet.transitionCount(), "XOR with 2 branches should expand to 2 flat transitions");

        var ft0 = flatNet.transitions().get(0);
        var ft1 = flatNet.transitions().get(1);

        // Both should consume from input
        int inIdx = flatNet.indexOf(input);
        assertEquals(1, ft0.preVector()[inIdx]);
        assertEquals(1, ft1.preVector()[inIdx]);

        // Each should produce to different output
        int out1Idx = flatNet.indexOf(out1);
        int out2Idx = flatNet.indexOf(out2);

        // One produces to out1, other to out2
        boolean branch0HasOut1 = ft0.postVector()[out1Idx] == 1;
        boolean branch1HasOut1 = ft1.postVector()[out1Idx] == 1;
        assertTrue(branch0HasOut1 ^ branch1HasOut1, "Exactly one branch should output to Out1");
    }
}

File: typescript/tests/verification/net-flattener.test.ts
import { describe, it, expect } from 'vitest';
import { flatten, bounded } from '../../src/verification/encoding/net-flattener.js';
import { PetriNet } from '../../src/core/petri-net.js';
import { Transition } from '../../src/core/transition.js';
import { place, environmentPlace } from '../../src/core/place.js';
import { one, exactly, all, atLeast } from '../../src/core/in.js';
import { outOne, outExactly, and, xor, andPlaces, xorPlaces } from '../../src/core/out.js';

describe('NetFlattener', () => {
  it('assigns stable place indices sorted by name', () => {
    const pC = place('C');
    const pA = place('A');
    const pB = place('B');
    const t = Transition.builder('T')
      .inputs(one(pC), one(pA))
      .outputs(outOne(pB))
      .build();
    const net = PetriNet.builder('N').transition(t).build();
    const flatNet = flatten(net);

    // Places should be sorted by name: A, B, C
    expect(flatNet.places[0]!.name).toBe('A');
    expect(flatNet.places[1]!.name).toBe('B');
    expect(flatNet.places[2]!.name).toBe('C');

    // Reverse lookup consistent
    for (let i = 0; i < flatNet.places.length; i++) {
      expect(flatNet.placeIndex.get(flatNet.places[i]!.name)).toBe(i);
    }
  });

  it('flattenTransition pre-vector correct for one()', () => {
    const input = place('IN');
    const output = place('OUT');
    const t = Transition.builder('T')
      .inputs(one(input))
      .outputs(outOne(output))
      .build();
    const net = PetriNet.builder('N').transition(t).build();
    const flatNet = flatten(net);

    expect(flatNet.transitions).toHaveLength(1);
    const ft = flatNet.transitions[0]!;
    const inIdx = flatNet.placeIndex.get('IN')!;
    const outIdx = flatNet.placeIndex.get('OUT')!;

    expect(ft.preVector[inIdx]).toBe(1);
    expect(ft.postVector[outIdx]).toBe(1);
    expect(ft.preVector[outIdx]).toBe(0);
    expect(ft.postVector[inIdx]).toBe(0);
  });

  it('flattenTransition pre-vector correct for exactly()', () => {
    const input = place('IN');
    const output = place('OUT');
    const t = Transition.builder('T')
      .inputs(exactly(3, input))
      .outputs(outOne(output))
      .build();
    const net = PetriNet.builder('N').transition(t).build();
    const flatNet = flatten(net);
    const ft = flatNet.transitions[0]!;
    const inIdx = flatNet.placeIndex.get('IN')!;
    expect(ft.preVector[inIdx]).toBe(3);
  });

  it('flattenTransition consumeAll flag for all()', () => {
    const input = place('IN');
    const output = place('OUT');
    const t = Transition.builder('T')
      .inputs(all(input))
      .outputs(outOne(output))
      .build();
    const net = PetriNet.builder('N').transition(t).build();
    const flatNet = flatten(net);
    const ft = flatNet.transitions[0]!;
    const inIdx = flatNet.placeIndex.get('IN')!;
    expect(ft.preVector[inIdx]).toBe(1);
    expect(ft.consumeAll[inIdx]).toBe(true);
  });

  it('flattenTransition consumeAll flag for atLeast()', () => {
    const input = place('IN');
    const output = place('OUT');
    const t = Transition.builder('T')
      .inputs(atLeast(3, input))
      .outputs(outOne(output))
      .build();
    const net = PetriNet.builder('N').transition(t).build();
    const flatNet = flatten(net);
    const ft = flatNet.transitions[0]!;
    const inIdx = flatNet.placeIndex.get('IN')!;
    expect(ft.preVector[inIdx]).toBe(3);
    expect(ft.consumeAll[inIdx]).toBe(true);
  });

  it('flattenTransition records inhibitor and read arcs', () => {
    const input = place('IN');
    const output = place('OUT');
    const inh = place('INH');
    const rd = place('RD');
    const t = Transition.builder('T')
      .inputs(one(input))
      .outputs(outOne(output))
      .inhibitor(inh)
      .read(rd)
      .build();
    const net = PetriNet.builder('N').transition(t).build();
    const flatNet = flatten(net);
    const ft = flatNet.transitions[0]!;

    expect(ft.inhibitorPlaces).toContain(flatNet.placeIndex.get('INH'));
    expect(ft.readPlaces).toContain(flatNet.placeIndex.get('RD'));
  });

  it('flattenTransition records reset arcs', () => {
    const input = place('IN');
    const output = place('OUT');
    const rst = place('RST');
    const t = Transition.builder('T')
      .inputs(one(input))
      .outputs(outOne(output))
      .reset(rst)
      .build();
    const net = PetriNet.builder('N').transition(t).build();
    const flatNet = flatten(net);
    const ft = flatNet.transitions[0]!;
    expect(ft.resetPlaces).toContain(flatNet.placeIndex.get('RST'));
  });

  it('XOR outputs expand to separate flat transitions', () => {
    const input = place('IN');
    const outA = place('A');
    const outB = place('B');
    const t = Transition.builder('T')
      .inputs(one(input))
      .outputs(xorPlaces(outA, outB))
      .build();
    const net = PetriNet.builder('N').transition(t).build();
    const flatNet = flatten(net);

    expect(flatNet.transitions).toHaveLength(2);
    expect(flatNet.transitions[0]!.name).toBe('T_b0');
    expect(flatNet.transitions[1]!.name).toBe('T_b1');
    expect(flatNet.transitions[0]!.branchIndex).toBe(0);
    expect(flatNet.transitions[1]!.branchIndex).toBe(1);

    // Each branch should produce to exactly one output
    const idxA = flatNet.placeIndex.get('A')!;
    const idxB = flatNet.placeIndex.get('B')!;
    expect(flatNet.transitions[0]!.postVector[idxA]).toBe(1);
    expect(flatNet.transitions[0]!.postVector[idxB]).toBe(0);
    expect(flatNet.transitions[1]!.postVector[idxA]).toBe(0);
    expect(flatNet.transitions[1]!.postVector[idxB]).toBe(1);
  });

  it('non-XOR transition maps 1:1', () => {
    const input = place('IN');
    const output = place('OUT');
    const t = Transition.builder('T')
      .inputs(one(input))
      .outputs(outOne(output))
      .build();
    const net = PetriNet.builder('N').transition(t).build();
    const flatNet = flatten(net);

    expect(flatNet.transitions).toHaveLength(1);
    expect(flatNet.transitions[0]!.name).toBe('T');
    expect(flatNet.transitions[0]!.branchIndex).toBe(-1);
  });

  it('AND outputs produce to all places', () => {
    const input = place('IN');
    const outA = place('A');
    const outB = place('B');
    const t = Transition.builder('T')
      .inputs(one(input))
      .outputs(andPlaces(outA, outB))
      .build();
    const net = PetriNet.builder('N').transition(t).build();
    const flatNet = flatten(net);

    expect(flatNet.transitions).toHaveLength(1);
    const ft = flatNet.transitions[0]!;
    expect(ft.postVector[flatNet.placeIndex.get('A')!]).toBe(1);
    expect(ft.postVector[flatNet.placeIndex.get('B')!]).toBe(1);
  });

  it('sink transition with no outputs has empty post-vector', () => {
    const input = place('IN');
    const t = Transition.builder('T')
      .inputs(one(input))
      .build();
    const net = PetriNet.builder('N').transition(t).build();
    const flatNet = flatten(net);

    expect(flatNet.transitions).toHaveLength(1);
    const ft = flatNet.transitions[0]!;
    expect(ft.postVector.every(v => v === 0)).toBe(true);
  });

  it('environment bounds applied in bounded mode', () => {
    const envP = environmentPlace<string>('ENV');
    const input = envP.place;
    const output = place('OUT');
    const t = Transition.builder('T')
      .inputs(one(input))
      .outputs(outOne(output))
      .build();
    const net = PetriNet.builder('N').transition(t).build();
    const flatNet = flatten(net, new Set([envP]), bounded(3));

    expect(flatNet.environmentBounds.get('ENV')).toBe(3);
  });

  it('multiple transitions flatten correctly', () => {
    const p1 = place('P1');
    const p2 = place('P2');
    const p3 = place('P3');

    const t1 = Transition.builder('T1')
      .inputs(one(p1))
      .outputs(outOne(p2))
      .build();
    const t2 = Transition.builder('T2')
      .inputs(one(p2))
      .outputs(outOne(p3))
      .build();

    const net = PetriNet.builder('Pipeline').transitions(t1, t2).build();
    const flatNet = flatten(net);

    expect(flatNet.places).toHaveLength(3);
    expect(flatNet.transitions).toHaveLength(2);
  });

  // ==================== Multiplicity (IO-018, IO-019) ====================

  it('Out.and(P, P, P) post-vector reflects count 3', () => {
    const input = place('Input');
    const output = place('Output');
    const t = Transition.builder('Triple')
      .inputs(one(input))
      .outputs(andPlaces(output, output, output))
      .build();
    const net = PetriNet.builder('RepeatedAndNet').transition(t).build();
    const flatNet = flatten(net);

    expect(flatNet.transitions).toHaveLength(1);
    const ft = flatNet.transitions[0]!;
    const outIdx = flatNet.placeIndex.get(output.name)!;
    expect(ft.postVector[outIdx]).toBe(3);
  });

  it('Out.exactly(5, P) post-vector equals count', () => {
    const input = place('Input');
    const output = place('Output');
    const t = Transition.builder('Five')
      .inputs(one(input))
      .outputs(outExactly(5, output))
      .build();
    const net = PetriNet.builder('ExactlyNet').transition(t).build();
    const flatNet = flatten(net);

    const ft = flatNet.transitions[0]!;
    const outIdx = flatNet.placeIndex.get(output.name)!;
    expect(ft.postVector[outIdx]).toBe(5);
  });

  it('AND of two exactly accumulates counts', () => {
    const input = place('Input');
    const output = place('Output');
    const t = Transition.builder('Sum')
      .inputs(one(input))
      .outputs(and(outExactly(2, output), outExactly(3, output)))
      .build();
    const net = PetriNet.builder('SumNet').transition(t).build();
    const flatNet = flatten(net);

    const ft = flatNet.transitions[0]!;
    const outIdx = flatNet.placeIndex.get(output.name)!;
    expect(ft.postVector[outIdx]).toBe(5);
  });

  it('XOR with different counts: each branch carries its own count', () => {
    const input = place('Input');
    const output = place('Output');
    const t = Transition.builder('Choice')
      .inputs(one(input))
      .outputs(xor(outOne(output), outExactly(3, output)))
      .build();
    const net = PetriNet.builder('XorCountsNet').transition(t).build();
    const flatNet = flatten(net);

    expect(flatNet.transitions).toHaveLength(2);
    const outIdx = flatNet.placeIndex.get(output.name)!;
    expect(flatNet.transitions[0]!.postVector[outIdx]).toBe(1);
    expect(flatNet.transitions[1]!.postVector[outIdx]).toBe(3);
  });
});
